app/core/utils.py

```diff
@@ -197,3 +202,145 @@ def get_gpu_info():
     mem_percentage = (used_memory_value / total_memory_value) * 100
 
     return gpu_name, total_memory_value, used_memory_value, mem_percentage
+
+
+def extract_labels_from_html_file(html_content, class_names):
+    soup = BeautifulSoup(html_content, "html.parser")
+    labels = soup.select(class_names)
+    return (label.get_text() for label in labels)
+
+
+def find_maximum_label(labels):
+    pattern = re.compile(r"v\d+\.\d+\.\d+$")
+    return max(filter(pattern.match, labels), default=None)
+
+
+def get_premd_last_tag(owner, repository, package):
+    response = requests.get(
+        f"https://github.com/{owner}/{repository}/pkgs/container/{package}"
+    )
+    try:
+        labels = extract_labels_from_html_file(
+            response.content, ".Label.mr-1.mb-2.text-normal"
+        )
+    except Exception as e:
+        logger.info(f"Unexpected error: {e}")
+    else:
+        return find_maximum_label(labels)
+
+
+def get_local_docker_image_tags(owner, repository):
+    try:
+        client = get_docker_client()
+        image = client.images.get(f"ghcr.io/{owner}/{repository}")
+        return image.tags
+    except Exception as e:
+        logger.info(f"Unexpected error: {e}")
+        return []
+
+
+def generate_container_name(prefix):
+    client = get_docker_client()
+
+    containers = client.containers.list(
+        all=True, filters={"name": f"^{prefix}", "status": "running"}
+    )
+    latest_suffix = -1
+    for container in containers:
+        match = re.match(rf"{prefix}_(\d+)", container.name)
+        if match:
+            suffix = int(match.group(1))
+            if suffix > latest_suffix:
+                latest_suffix = suffix
+
+    if latest_suffix == -1:
+        return prefix, f"{prefix}_1"
+    else:
+        return f"{prefix}_{latest_suffix}", f"{prefix}_{latest_suffix+1}"
```

Comment:
shouldn't be needed at all (just generate an auto-named `new_container`, then remove the old `premd` container, then `new_container.rename("premd")`)

```suggestion
```